Fix issue with "to_date" failing to process dates later than year 2262

MartinKolbAtWork opened the pull request:
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/12226

## Rationale for this change

The "to_date" function (https://github.com/apache/datafusion/blob/main/datafusion/functions/src/datetime/to_date.rs) fails to process dates that are later than year 2262.
This is caused by the implementation detail that the conversion process uses nano-seconds based on epoch.
The Arrow datatype for Date32 and Date64 support much larger values. The statements in some areas in the code state that the usage of nanoseconds is imposed by the Date types of Arrow. This is simply wrong. The Date32 type stores the number of days since epoch. The Date64 type stores the milliseconds (NOT nanoseconds) since epoch. Both Date32 and Date64 can therefore massively exceed the year 2262.
See: https://arrow.apache.org/docs/cpp/api/datatype.html

NOTE:
Processing dates later than 2262 is not a theoretical issue. In widely used business software systems, unbounded dates (e.g. an "expiry_date" that is set to never expire) are set to the 31st of December of the year 9999 (i.e. "9999-12-31").
Processing such data with datafusion will fail if the current "to_date" implementation touches this data.

## What changes are included in this PR?
- Changing the implementation of "to_date" to use milli-seconds instead of nano-seconds in order to achieve compliance with the values supported on Arrow's Date32 and Date64
- Unit tests to validate the correct behavior

## Are these changes tested?

Unit tests are provided in https://github.com/apache/datafusion/blob/main/datafusion/functions/src/datetime/to_date.rs

## Are there any user-facing changes?

n/a

comphead reviewed (commented):
Thanks @MartinKolbAtWork for your contribution

I just checked we do have an issue with `to_date` function whereas the `cast` works correctly

```
> select extract(day from cast('9999-12-31' as date));
+-------------------------------------------+
| date_part(Utf8("DAY"),Utf8("9999-12-31")) |
+-------------------------------------------+
| 31.0                                      |
+-------------------------------------------+
1 row(s) fetched. 
Elapsed 0.005 seconds.

> select to_date('9999-12-31');
Arrow error: Parser error: The dates that can be represented as nanoseconds have to be between 1677-09-21T00:12:44.0 and 2262-04-11T23:47:16.854775804
```

Perhaps we can use a the same implementation like CAST does(reuse it) instead of creating a new impl

findepi commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -118,3 +134,52 @@ impl ScalarUDFImpl for ToDateFunc {
+
+    #[test]
+    fn test_year_9999() {
+        let date_str = "9999-12-31";
```
this is a good date to test with

can you also add a test with other interesting dates?

- year 0001 (also common in some data, just like 9999)
- a year before 1970 epoch
- dates during julian/gregorian switch for some western countries (1582-10-10)
- negative years -- these represent problems to some parsing libraries
- 5+-digit year (after 9999) -- these represent problems to some parsing libraries

MartinKolbAtWork replied:
Thanks for the suggestions. I added tests for (almost) all the examples you mentioned. To make negative years work, I had to adapt the code, because currently negative years also do not work.
I intentionally did not add support for 5-digit years, because not even the ISO 8601 standard supports them.

findepi replied:
i have no issue not supporting 5+-digit years, but interestingly they also can show in real data. imagine some data using 9999-12-31 as a placeholder and then this being converted to a timestamp with time zone and then converted to a different zone... this may produce 10000-01-01. And indeed, at least some query engines do support such dates, but you're right that 9999-12-31 is what they "are obliged" to support.

just saying... it's fine to omit this here for now


>  I added tests for (almost) all the examples you mentioned.


thank you!

MartinKolbAtWork commented:
> Thanks @MartinKolbAtWork for your contribution
> 
> I just checked we do have an issue with `to_date` function whereas the `cast` works correctly
> 
> ```
> > select extract(day from cast('9999-12-31' as date));
> +-------------------------------------------+
> | date_part(Utf8("DAY"),Utf8("9999-12-31")) |
> +-------------------------------------------+
> | 31.0                                      |
> +-------------------------------------------+
> 1 row(s) fetched. 
> Elapsed 0.005 seconds.
> 
> > select to_date('9999-12-31');
> Arrow error: Parser error: The dates that can be represented as nanoseconds have to be between 1677-09-21T00:12:44.0 and 2262-04-11T23:47:16.854775804
> ```
> 
> Perhaps we can use a the same implementation like CAST does(reuse it) instead of creating a new impl

The main thing of "to_date" is that is supports supplying a format string. As CAST does not support this, I don't think merging or re-using these two implementations is an approach that should be pursued.

findepi commented:
@MartinKolbAtWork please check the build results.
for example

```
[Diff] (-expected|+actual)
-   2023-01-13
+   2023-01-14
at test_files/dates.slt:126
```

it might be the expected value was wrong in this case

findepi commented:
Can you please update the docs?

https://github.com/apache/datafusion/blob/b691b35f6904942c3544ae2af80aef927dfc8d11/docs/source/user-guide/sql/scalar_functions.md#L1739

comphead commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -47,27 +48,24 @@ impl ToDateFunc {
-                        })
+                |s, format| match Date32Type::parse_formatted(s, format) {
+                    Some(v) => Ok(v),
+                    None => Err(internal_datafusion_err!(
```
```suggestion
                    None => internal_err!(
```

MartinKolbAtWork replied:
Code was updated and partially re-structured.

comphead commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -47,27 +48,24 @@ impl ToDateFunc {
+                // values for year (e.g. years in BC). Date32Type::parse_formatted can handle also these.
+                |s| match Date32Type::parse_formatted(s, "%Y-%m-%d") {
+                    Some(v) => Ok(v),
+                    None => Err(internal_datafusion_err!(
```
```suggestion
                    None => internal_err!(
```

comphead commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -47,27 +48,24 @@ impl ToDateFunc {
+                // For most cases 'Date32Type::parse' would also work here, as it semantically parses
+                // assuming a YYYY-MM-DD format. However 'Date32Type::parse' cannot handle negative
+                // values for year (e.g. years in BC). Date32Type::parse_formatted can handle also these.
+                |s| match Date32Type::parse_formatted(s, "%Y-%m-%d") {
```
I'm not sure the string format will be exactly that format?

comphead replied:
I just checked DuckDB so it expects the format provided 
```
select extract(day from cast('9999-31-12' as date));
Error: Conversion Error: date field value out of range: "9999-31-12", expected format is (YYYY-MM-DD)
```

MartinKolbAtWork commented:
Hi @findepi I fixed the test failures and updated the docs. You the build be re-triggered to check if this was successful?

findepi commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -47,22 +57,39 @@ impl ToDateFunc {
             1 => handle::<Date32Type, _, Date32Type>(
                 args,
+                // Although not being speficied, previous implementations have supported parsing date values that are followed
+                // by a time part separated either a whitespace ('2023-01-10 12:34:56.000') or a 'T' ('2023-01-10T12:34:56.000').
```
- Was the previous implementation accepting `<date>T<time>` or `<date>T<anything>`? The comment suggests the former, the code now accepts the latter
- if this is desirable behavior, we should document it (and then we can use a different code comment wording -- not referring to previous implementation as the one constraining us for ever). If this isn't desirable behavior, we should fix it. In any case it might be best to address this as a follow-up issue. @comphead WDYT?

findepi commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -47,22 +57,39 @@ impl ToDateFunc {
+                        string_to_parse = s.split_whitespace().next();
+                    }
+                    if string_to_parse.is_none() {
+                        return Err(internal_datafusion_err!(
```
please try to use `internal_err!` per https://github.com/apache/datafusion/pull/12227#discussion_r1738902790

findepi commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -47,22 +57,39 @@ impl ToDateFunc {
+                    // assuming a YYYY-MM-DD format. However 'Date32Type::parse' cannot handle negative
+                    // values for year (e.g. years in BC). Date32Type::parse_formatted can handle also these.
+                    match Date32Type::parse_formatted(string_to_parse.unwrap(), "%Y-%m-%d") {
+                    Some(v) => Ok(v),
```
formatting seems off? typically `match` options are indented 
(`cargo fmt` seems to be fine either way, not sure why)

findepi reviewed (commented):
thanks for the update! couple minor comments

MartinKolbAtWork commented:
Hi @findepi ,
thanks again for assisting me to get the PR in a “mergeable” state. Really appreciate your input. Your comment on the change regarding the timestamp behavior made me think that we might take one step back.
These changes related to the timestamp handling were added to support negative years. The initial target for this PR is fixing the problem of handling years greater than 2262. I found your suggestion about supporting negative years very compelling, because it semantically addresses a similar edge case.

Now that we found out that adding support for negative years does add several nitty-gritty details, that need to be carefully thought through, I suggest that we split this up and handle the negative years topic in a separate PR.
I reverted the changes back to the point where it’s only about supporting years up to 9999. This makes the code simpler and easier to understand.
Hope this makes sense for you.

alamb commented:
fyi @Omega359

Omega359 commented:
I wonder what the cost is for using string_to_datetime_formatted vs a more limited string_to_date_formatted. There should be slightly less parsing involved for that.

MartinKolbAtWork commented:
> I wonder what the cost is for using string_to_datetime_formatted vs a more limited string_to_date_formatted. There should be slightly less parsing involved for that.

Hi @Omega359 ,
thanks for asking. This is a good question. Let me throw in two aspects.

1) The current implementation also uses a “complete” parsing, so the proposed change does not make anything worse. The change just uses milliseconds instead of nanoseconds as an intermediate value to avoid the limitation of the upper bound of year 2262 when using nanoseconds.

2) Simply parsing the values for year, month, and day out of a timestamp will not always create correct value for a date. This is an example of an existing test case in DataFusion: https://github.com/apache/datafusion/blob/dd3208943d728d845497d6a12ce4c0eacc061dcd/datafusion/sqllogictest/test_files/dates.slt#L127-L129
The correct day of the timestamp `01-14-2023 01:01:30+05:30` is `13`, although the “parsed day” is `14`. This is because the timestamp refers to a time-zone where the local day is already the 14th, but the UTC day is the 13th.
When parsing to a “timestamp”, it is expected that the “day” is “14” because the timestamp object can be asked for its time-zone, so the UTC date (13th) can be calculated from that available information. A simple “date” object however does not carry time-zone information. Therefore the date must refer to UTC, which is the 13th.

And, as a brief reminder… I just intended to fix the issue that the date calculation has an upper limit of 2262. Both, the full timestamp parsing as well as the UTC handling are already in the code-base. I introduced neither of them.

Omega359 commented:
> And, as a brief reminder… I just intended to fix the issue that the date calculation has an upper limit of 2262. Both, the full timestamp parsing as well as the UTC handling are already in the code-base. I introduced neither of them.

Good points. I am unsure if I introduced that or just inherited it when I added format parsing and didn't think closely enough about it. Glad to see this being resolved

MartinKolbAtWork commented:
Hi @Omega359 , @findepi , @alamb ,
thanks for your suggestions for far. I'm not sure if there anything is left that you want me to change, or do we consider the current state to be ready for getting an approving review?

alamb commented:
> thanks for your suggestions for far. I'm not sure if there anything is left that you want me to change, or do we consider the current state to be ready for getting an approving review?

I will check this PR out shortly

MartinKolbAtWork commented:
> > thanks for your suggestions for far. I'm not sure if there anything is left that you want me to change, or do we consider the current state to be ready for getting an approving review?
> 
> I will check this PR out shortly

@alamb , take your time.
I just asked to make sure that I did not miss any action that someone is waiting for.

MartinKolbAtWork commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -36,6 +39,13 @@ impl Default for ToDateFunc {
     }
 }
 
+fn string_to_timestamp_millis_formatted(s: &str, format: &str) -> Result<i64> {
```
OK. I moved it with the latest commit.

alamb commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -118,3 +118,212 @@ impl ScalarUDFImpl for ToDateFunc {
+
+    use super::ToDateFunc;
+
+    #[test]
```
😍 

Very nice

alamb reviewed (approved):
Looks really nice (and well tested to me). THank you @MartinKolbAtWork 

Also, thank you @findepi  and @comphead for the review

cc @Omega359

alamb commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -118,3 +118,212 @@ impl ScalarUDFImpl for ToDateFunc {
+            "2020-09-08 12:13:29",
+        ];
+        for date_str in test_cases {
+            let formatted_date_scalar = ScalarValue::Utf8(Some(date_str.into()));
```
BTW I think you can write the same thing more concisely like 

```suggestion
            let formatted_date_scalar = ScalarValue::from(date_str);
```

alamb commented on datafusion/functions/src/datetime/to_date.rs:
```diff
@@ -118,3 +118,212 @@ impl ScalarUDFImpl for ToDateFunc {
+        for date_str in test_cases {
+            let formatted_date_scalar = ScalarValue::Utf8(Some(date_str.into()));
+
+            let to_date_result =
+                ToDateFunc::new().invoke(&[ColumnarValue::Scalar(formatted_date_scalar)]);
+
+            match to_date_result {
+                Ok(ColumnarValue::Scalar(ScalarValue::Date32(date_val))) => {
```
The machinery to call the function is also a but clunky, I wonder if it would make the tests easier to read (as there would be less conversion machinery) if we added a function like

```rust
/// Invokes ToDate on the input string, and returns the resulting Date32 value
fn run_to_date(input: &str) -> Result<i32> {
..
}
```

alamb commented:
I believe the CI failure  https://github.com/apache/datafusion/actions/runs/10725200169/job/29798407173?pr=12227 was fixed on main by @mbrobbel  in https://github.com/apache/datafusion/pull/12346

I took the liberty of merging the branch up from main to try and get a clean CI run

alamb commented:
We can consider potential changes to make the test more "readable" as part of a follow on PR.

Thanks agian @MartinKolbAtWork and @findepi
